### ISSUE TITLE
Cuckoo filter in Design2: questions and assumptions

### ISSUE THREAD

**Issue by dan-blank** [DP-3T/documents#142]
I am writing a simulation to quickly try out different combinations of data structures for storing EphIDs, mainly combinations of lists, cuckoo filters and blooms filters with varying degrees of false positive rates (from now FPP). I want to analyze how many people in an population with a given rate of infection had to downlad what amount of data per download. My hope is that a combination such as "Very low FPP cuckoo filter + very high FPP bloom filter" will reduce of transfered data considerably. 
There are a few things I did not understand yet - I would be grateful for answers if you could help me there :)

1) As I understood, the size of a stored item in a cuckoo filter does not matter. So currently you store the 32 byte seed not because it is smaller than the 48 byte EphID, but because you need to store the seed only once per infected patient instead of all EphIDs generated in the infectous time?

2) In the worst case (2 weeks of history), a patient needs to upload 1344 EphIDs (4 new EphIDs per hours, 24 hours, 14 days), correct?

3) What is an acceptable FPP (= such that a user can be confident with an answer) for you?

4) According to http://www.eecs.harvard.edu/~michaelm/postscripts/cuckoo-conext2014.pdf, the bits needed to store one entry in a cuckoo filter can be calculated by [ log2(1/FPP) + log2(2b) ] / a. What is the configuration of a and b that you use / works well?

5) Regarding the observation in Design 2, Scalability: "For 40.000 new infections per day (approximately the cumulative total of 5 European countries at their contagion peak) smartphones then need to download 100 MB each day." I think the 100MB are the result of 40.000 seeds stored in a cuckoo filter with some b (2 <= b <= 4) and a very low a. Is this correct and what are the exact values of a and b? (I am asking because I want to make sure to hit the same numbers as you do in my simulation).

**Comment by dirkx** [comment]
@dan-blank  I thought along the same lines - and try to capture what would need to be defined to have a interoperable EU serialized CF over here https://github.com/dirkx/DP-3T-Documents/blob/implementation-profile-start/implementation-profiles/profile.md - with running code that does it at https://github.com/dirkx/DP-3T-Documents/tree/editable-version/impl/design-2-openssl-C.

Would be interesting to compare notes with your serialisation.

**Comment by dan-blank** [comment]
@dirkx Thank you very much! The links were exactly what I needed. What I have in mind is a rather high-level simulation to quickly explore the design space of using multiple data structures, relying on rough assumptions about space usage.

That should take me 2 days. After that I see if I can contribute further with a concrete serialisation for example. If I find anything interesting, I plan to report it in a new issue with a more fitting title.

**Comment by dirkx** [comment]
So I tried to make the profile such that one can optimize the backend services without changing the apps in the field.

If you find that going from bytes to bits is worth it (rather than focus on beter packing) then that would be good to know early.
